### PR TITLE
Add recipe for openslr-85 dataset

### PR DESCRIPTION
**Pull request by nwpuaslp**
We add a recipe of keyword spotting for openslr-85 data set. In our script we provide two method. One is using chain model which is reference from mobvoi, the other is using keyword-filler. There are two decoding methods: posterior smoothing and FST decoding.

### CODE REVIEW

**Comment by danpovey** [comment]
Great, @freewym can you please review?

**Comment by freewym** [comment]
OK

File: egs/hi_mia/v1/RESULT [hunk 1]
```diff
@@ -0,0 +1,6 @@
+run_kws_kf.sh
+shreld0.5
+accuracy 0.83273214201
```
Comment (csukuangfj) [inline egs/hi_mia/v1/RESULT:3]: Could you please provide false alarm per hour and  recall?
These two metrics are more important in actual applications.

**Comment by nwpuaslp** [comment]
Ok, I will provide it soon

发送自 Windows 10 版邮件<https://go.microsoft.com/fwlink/?LinkId=550986>应用

________________________________
发件人: Fangjun Kuang <notifications@github.com>
发送时间: Tuesday, May 12, 2020 4:13:46 PM
收件人: kaldi-asr/kaldi <kaldi@noreply.github.com>
抄送: nwpuaslp <github-nwpuaslp@outlook.com>; Author <author@noreply.github.com>
主题: Re: [kaldi-asr/kaldi] Add recipe for openslr-85 dataset (#4064)


@csukuangfj commented on this pull request.

________________________________

In egs/hi_mia/v1/RESULT<https://github.com/kaldi-asr/kaldi/pull/4064#discussion_r423545359>:

> @@ -0,0 +1,6 @@
+run_kws_kf.sh
+shreld0.5
+accuracy 0.83273214201


Could you please provide false alarm per hour and recall?
These two metrics are more important in actual applications.

―

File: egs/hi_mia/v1/local/kws_draw_roc.py [hunk 2]
```diff
@@ -0,0 +1,70 @@
+#!/usr/bin/env python
```
Comment (freewym) [inline egs/hi_mia/v1/local/kws_draw_roc.py:1]: #!/usr/bin/env python3

File: egs/hi_mia/v1/local/kws_draw_roc.py [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+    x = list()
+    y = list()
+    for i in range(1,1001):
+        false_reject=0
+        false_alarm=0
+        positive=0
+        negative=0
+        true_alarm=0
+        true_reject=0
+        print "shreld"+str(float(i)/1000)
```
Comment (freewym) [inline egs/hi_mia/v1/local/kws_draw_roc.py:42]: use print("....") for python3, here and any other places

File: egs/hi_mia/v1/local/kws_draw_roc.py [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+        for key in resultDic:
+            label = int(labelDic[key])
+            result = float(resultDic[key])
+            if label ==1:
```
Comment (freewym) [inline egs/hi_mia/v1/local/kws_draw_roc.py:46]: add a single space before and after each operator; here and any other places

File: egs/hi_mia/v1/local/kws_posterior_handling.py [hunk 5]
```diff
@@ -0,0 +1,72 @@
+#!/usr/bin/env python
```
Comment (freewym) [inline egs/hi_mia/v1/local/kws_posterior_handling.py:1]: make changes as in other python files.

File: egs/hi_mia/v1/local/nnet3/tuning/run_tdnn_1a.sh [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+# training options
+initial_effective_lrate=0.0015
+final_effective_lrate=0.00015
+num_epochs=2
+num_jobs_initial=1
+num_jobs_final=1
+nj=30
+remove_egs=true
+num_targets=
+. parse_options.sh || exit 1;
```
Comment (freewym) [inline egs/hi_mia/v1/local/nnet3/tuning/run_tdnn_1a.sh:26]: remove this line.

File: egs/hi_mia/v1/local/nnet3/tuning/run_tdnn_1a.sh [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+fi
+
+# we use 40-dim high-resolution mfcc features (w/o pitch and ivector) for nn training
+# no utt- and spk- level cmvn
+
+dir=exp/nnet3/tdnn_test${affix:+_$affix}
+gmm_dir=exp/tri3
+test_sets="fbank/dev fbank/test"
+train_set=fbank/train
+ali_dir=/home/work_nfs3/zhyyao/yao_workspace/kws_mia/exp/kws_ali
```
Comment (freewym) [inline egs/hi_mia/v1/local/nnet3/tuning/run_tdnn_1a.sh:52]: change the absolute path

File: egs/hi_mia/v1/local/process_result.py [hunk 8]
```diff
@@ -0,0 +1,47 @@
+#!/usr/bin/env python
```
Comment (freewym) [inline egs/hi_mia/v1/local/process_result.py:1]: make changes as in other python files.

File: egs/hi_mia/v1/local/run_fst.py [hunk 9]
```diff
@@ -0,0 +1,215 @@
+#!/usr/bin/env python
```
Comment (freewym) [inline egs/hi_mia/v1/local/run_fst.py:1]: make changes as in other python files.

File: egs/hi_mia/v1/local/run_fst.py [hunk 10]
```diff
@@ -0,0 +1,215 @@
...
+#           2020 Beijing Shell Shell Tech. Co. Ltd. (Author: Hui BU) 
+# Apache 2.0
+
+# This script is for keyword filler system to do fst
+
+import numpy as np
+import argparse
+import sys
+import copy
+class edge:
```
Comment (freewym) [inline egs/hi_mia/v1/local/run_fst.py:16]: First letter of class names should be capitalized

File: egs/hi_mia/v1/local/run_fst.py [hunk 11]
```diff
@@ -0,0 +1,215 @@
...
+                    self.max_score_of_current_state = score
+                    self.average_max_keyword_score_before = prev.average_max_keyword_score
+                self.average_keyword_score=(self.max_score_of_current_state + 
+                    self.average_max_keyword_score_before * (self.num_keyword_states - 1)) / self.num_keyword_states
+                if olabel != 0:
+                    self.keyword=olabel
+            self.active=True
+            self.is_filler= is_filler
+                
+    def Reset():
```
Comment (freewym) [inline egs/hi_mia/v1/local/run_fst.py:85]: naming convention for class methods should be consistent with others, e.g. lower case for the first letter

File: egs/hi_mia/v1/local/run_fst.py [hunk 12]
```diff
@@ -0,0 +1,215 @@
...
+    parser.add_argument('bnf_file',help='output of network')
+    
+    FLAGS = parser.parse_args()
+
+    stateM = stateMachine(FLAGS.Fst_file)
+    result = stateM.runFst(FLAGS.bnf_file)
+
+    for k,v in result.items():
+        print str(k)+" "+str(v)            
+
```
Comment (freewym) [inline egs/hi_mia/v1/local/run_fst.py:214]: remove trailing new lines

File: egs/hi_mia/v1/local/run_fst.py [hunk 13]
```diff
@@ -0,0 +1,215 @@
...
+            return -1,-1
+        if pos == -1:
+            sys.exit(1)
+        return self.oedge[pos].to_state, self.oedge[pos].osymbol
+
+class token:
+    def __init__(self):
+        self.active=False
+        self.is_filler=True
+        self.score=0
```
Comment (csukuangfj) [inline egs/hi_mia/v1/local/run_fst.py:52]: There should be spaces before and after the assignment `=` operator.

**Comment by nwpuaslp** [comment]
I have fixed the problem in code and provide the result.

**Comment by danpovey** [comment]
Thanks!

**Comment by nwpuaslp** [comment]
Thanks for your efforts!

发送自 Windows 10 版邮件<https://go.microsoft.com/fwlink/?LinkId=550986>应用

发件人: Daniel Povey<mailto:notifications@github.com>
发送时间: 2020年5月19日 23:30
收件人: kaldi-asr/kaldi<mailto:kaldi@noreply.github.com>
抄送: nwpuaslp<mailto:github-nwpuaslp@outlook.com>; Author<mailto:author@noreply.github.com>
主题: Re: [kaldi-asr/kaldi] Add recipe for openslr-85 dataset (#4064)


Merged #4064<https://github.com/kaldi-asr/kaldi/pull/4064> into master.

―

**Comment by framsc** [comment]
There are run_kws_chain.sh and run_kws_kf.sh in the recipe. What does _kf_ mean?

**Comment by framsc** [comment]
Is **kf** short for keyword-filler. I compare the differences between **run_kws_chain.sh** and **run_kws_kf.sh** . I can not figure out which part is the keyword-filler model.

**Comment by nwpuaslp** [comment]
Dear framsc,
Thanks for your question.
Yes,  kf is short for keyword-filler. We are aim to model  filler and keyword in a fst at phone level. You can see the code at run_fst.py. Hope this can help you.

发送自 Windows 10 版邮件<https://go.microsoft.com/fwlink/?LinkId=550986>应用

发件人: framsc<mailto:notifications@github.com>
发送时间: 2020年5月20日 20:12
收件人: kaldi-asr/kaldi<mailto:kaldi@noreply.github.com>
抄送: nwpuaslp<mailto:github-nwpuaslp@outlook.com>; Author<mailto:author@noreply.github.com>
主题: Re: [kaldi-asr/kaldi] Add recipe for openslr-85 dataset (#4064)


Is kf short for keyword-filler. I compare the differences between run_kws_chain.sh and run_kws_kf.sh . I can not figure out which part is the keyword-filler model.

―

**Comment by framsc** [comment]
@nwpuaslp Thank you very much. It solves my confusion.

**Comment by zhange1990** [comment]
I ran the script titled run_kws_chain.sh, and I got failure in the stage 10, where the code is as follows:
# train chain model
if [ $stage -le 10 ];then
    echo "stage 10"
    local/chain/tuning/run_tdnn_1a.sh
fi

and the infomation is as follows:
stage 10
local/chain/tuning/run_tdnn_1a.sh 
steps/nnet3/chain/build_tree.sh --frame-subsampling-factor 3 --context-opts --context-width=2 --central-position=1 --cmd run.pl 5000 data/fbank/train data/lang_chain exp/mono_ali exp/chain/mono_cd_tree_sp
steps/nnet3/chain/build_tree.sh: feature type is delta
steps/nnet3/chain/build_tree.sh: Initializing monophone model (for alignment conversion, in case topology changed)
steps/nnet3/chain/build_tree.sh: Accumulating tree stats
run.pl: 31 / 40 failed, log is in exp/chain/mono_cd_tree_sp/log/acc_tree.*.log

Then, the acc_tree.2.log is as follows:
 # convert-ali --frame-subsampling-factor=3 exp/mono_ali/final.mdl exp/chain/mono_cd_tree_sp/mono.mdl exp/chain/mono_cd_tree_sp/mo    no.tree "ark:gunzip -c exp/mono_ali/ali.2.gz|" ark:- | acc-tree-stats --context-width=2 --central-position=1 --ci-phones=1:2:3:4:    5 exp/chain/mono_cd_tree_sp/mono.mdl "ark,s,cs:apply-cmvn  --utt2spk=ark:data/fbank/train/split40/2/utt2spk scp:data/fbank/train/    split40/2/cmvn.scp scp:data/fbank/train/split40/2/feats.scp ark:- | add-deltas  ark:- ark:- | subsample-feats --n=3 ark:- ark:- |    " ark:- exp/chain/mono_cd_tree_sp/2.treeacc 
# Started at Fri May 29 01:06:03 CST 2020
#
convert-ali --frame-subsampling-factor=3 exp/mono_ali/final.mdl exp/chain/mono_cd_tree_sp/mono.mdl exp/chain/mono_cd_tree_sp/mono.tree 'ark:gunzip -c exp/mono_ali/ali.2.gz|' ark:-
acc-tree-stats --context-width=2 --central-position=1 --ci-phones=1:2:3:4:5 exp/chain/mono_cd_tree_sp/mono.mdl 'ark,s,cs:apply-cmvn  --utt2spk=ark:data/fbank/train/split40/2/utt2spk scp:data/fbank/train/split40/2/cmvn.scp scp:data/fbank/train/split40/2/feats.scp ark:- | add-deltas  ark:- ark:- | subsample-feats --n=3 ark:- ark:- |' ark:- exp/chain/mono_cd_tree_sp/2.treeacc
WARNING (convert-ali[5.5.532~1-7f0c9]:main():convert-ali.cc:89) Toplogies of models are not equal: conversion may not be correct or may fail.
WARNING (convert-ali[5.5.532~1-7f0c9]:ConvertAlignmentForPhone():hmm-utils.cc:765) Topology mismatch detected; automatically converting. Won't warn again.
apply-cmvn --utt2spk=ark:data/fbank/train/split40/2/utt2spk scp:data/fbank/train/split40/2/cmvn.scp scp:data/fbank/train/split40/2/feats.scp ark:-
add-deltas ark:- ark:-
subsample-feats --n=3 ark:- ark:-
LOG (convert-ali[5.5.532~1-7f0c9]:main():convert-ali.cc:120) Succeeded converting alignments for 5238 files, failed for 0
LOG (acc-tree-stats[5.5.532~1-7f0c9]:main():acc-tree-stats.cc:101) Processed 1000 utterances.
LOG (acc-tree-stats[5.5.532~1-7f0c9]:main():acc-tree-stats.cc:101) Processed 2000 utterances.
LOG (acc-tree-stats[5.5.532~1-7f0c9]:main():acc-tree-stats.cc:101) Processed 3000 utterances.
LOG (apply-cmvn[5.5.532~1-7f0c9]:main():apply-cmvn.cc:162) Applied cepstral mean normalization to 4229 utterances, errors on 0
LOG (subsample-feats[5.5.532~1-7f0c9]:main():subsample-feats.cc:115) Processed 4229 feature matrices; 0 with errors.
LOG (subsample-feats[5.5.532~1-7f0c9]:main():subsample-feats.cc:117) Processed 1980413 input frames and 661541 output frames.
LOG (acc-tree-stats[5.5.532~1-7f0c9]:main():acc-tree-stats.cc:118) Accumulated stats for 3167 files, 1062 failed due to no alignment, 0 failed for other reasons.

Since there were no failure in the up-stream stages, I wonder why the there are 1062 failed due to no alignment? 
By the way, there are the Warning that  the conversion of the alignment may not be correct or may fail, I wonder what the cause?

**Comment by danpovey** [comment]
That warning is expected.
You should show the error on one of the jobs that failed, that one succeeded.
Did you update your Kaldi repo in the middle of running this setup?  (May help diagnose the problem).
Likely something to do with data-splitting.
